src/main/scala/util/SynchronizerReg.scala

```diff
@@ -0,0 +1,202 @@
+}
+
+class AsyncResetSynchronizerShiftReg(w: Int = 1, sync: Int, init: Int)
+    extends AbstractPipelineReg(w) {
+  require(sync > 1, s"Sync must be greater than 1, not ${sync}.")
+  override def desiredName = s"AsyncResetSynchronizerShiftReg_w${w}_d${sync}_i${init}"
+  val output = Seq.tabulate(w) { i =>
+    val initBit = ((init >> i) & 1) > 0
+    SynchronizerPrimitiveShiftReg(io.d(i), sync, initBit,  SynchronizerResetType.Async)
+  }
+  io.q := Cat(output.reverse)
+}
+
+object AsyncResetSynchronizerShiftReg {
+  def apply [T <: Chisel.Data](in: T, sync: Int, init: Int, name: Option[String] = None): T =
+    AbstractPipelineReg(new AsyncResetSynchronizerShiftReg(in.getWidth, sync, init), in, name)
+
+  def apply [T <: Chisel.Data](in: T, sync: Int, name: Option[String]): T =
+    apply (in, sync, 0, name)
+
+  def apply [T <: Chisel.Data](in: T, sync: Int): T =
+    apply (in, sync, 0, None)
+
+  def apply [T <: Chisel.Data](in: T, sync: Int, init: T, name: Option[String]): T =
+    apply(in, sync, init.litValue.toInt, name)
+
+  def apply [T <: Chisel.Data](in: T, sync: Int, init: T): T =
+    apply (in, sync, init.litValue.toInt, None)
+}
+
+@deprecated("SyncResetSynchronizerShiftReg is unecessary with Chisel3 inferred resets. Use ResetSynchronizerShiftReg which will use the inferred reset type.", "rocket-chip 1.2")
+class SyncResetSynchronizerShiftReg(w: Int = 1, sync: Int, init: Int) extends AbstractPipelineReg(w) {
+  require(sync > 1, s"Sync must be greater than 1, not ${sync}.")
+  override def desiredName = s"SyncResetSynchronizerShiftReg_w${w}_d${sync}_i${init}"
+  val output = Seq.tabulate(w) { i =>
+    val initBit = ((init >> i) & 1) > 0
+    SynchronizerPrimitiveShiftReg(io.d(i), sync, initBit, SynchronizerResetType.Sync)
+  }
+  io.q := Cat(output.reverse)
+}
+
+object SyncResetSynchronizerShiftReg {
+  def apply [T <: Chisel.Data](in: T, sync: Int, init: Int, name: Option[String] = None): T =
+    if (sync == 0) in else AbstractPipelineReg(new SyncResetSynchronizerShiftReg(in.getWidth, sync, init), in, name)
+
+  def apply [T <: Chisel.Data](in: T, sync: Int, name: Option[String]): T =
+    apply (in, sync, 0, name)
+
+  def apply [T <: Chisel.Data](in: T, sync: Int): T =
+    apply (in, sync, 0, None)
+
+  def apply [T <: Chisel.Data](in: T, sync: Int, init: T, name: Option[String]): T =
+    apply(in, sync, init.litValue.toInt, name)
+
+  def apply [T <: Chisel.Data](in: T, sync: Int, init: T): T =
+    apply (in, sync, init.litValue.toInt, None)
+}
+
+class ResetSynchronizerShiftReg(w: Int = 1, sync: Int, init: Int) extends AbstractPipelineReg(w) {
+  require(sync > 1, s"Sync must be greater than 1, not ${sync}.")
+  override def desiredName = s"ResetSynchronizerShiftReg_w${w}_d${sync}_i${init}"
+  val output = Seq.tabulate(w) { i =>
+    val initBit = ((init >> i) & 1) > 0
+    SynchronizerPrimitiveShiftReg(io.d(i), sync, initBit, SynchronizerResetType.Inferred)
```

Comment:
It may've been cleaner to expose the reset type as a constructor argument to the class to avoid DRY on each of `SyncresetSynchronizerShiftReg`, `ResetSynchronzierShiftReg` since you do essentially the same thing in each of these classes `val output = Seq.tabulate(w) { ... }`, but that would've required a massive code refactor everywhere we currently instantiate these registers so I think this is ok for now